File: crates/agent/src/effect.rs
//! Effects (§11, GLOSSARY): serializable descriptions of something to be
//! done. The output of `step`; executed by the `core` runtime through ports;
//! results return as the next Event. Coarse on purpose (ARCHITECTURE §1c:
//! one CallModel, one InvokeTool — never micro-effects), so one Work turn is
//! one step in, one effect out, one event back.

use serde::{Deserialize, Serialize};

use context::{Document, ProviderFormat};
use kernel::{EndpointName, EventKind, ToolId};

/// The closed set of things an agent can ask the runtime to do — the §11
/// list, typed. Serializable because pending effects must survive a refresh
/// (replay reloads state + effects from the log, I11).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Effect {
    /// Call the model with an assembled paper (I13: nothing reaches a model
    /// except as a Document — this variant's type makes ad-hoc strings
    /// unrepresentable).
    CallModel {
        document: Document,
        format: ProviderFormat,
        endpoint: EndpointName,
        /// The agent's `model:` catalogue key (increment 04). Symbolic: the
        /// adapter resolves it against `public/models.json`, so no URL and no
        /// concrete model id exists anywhere upstream of the broker (I6).
        model: String,
        /// The agent's `temperature:` frontmatter key, or `None` where the
        /// file named none. It rides the effect rather than being read from a
        /// setting because it is a property of the AGENT, and the summarizer's
        /// turn is a different agent's call inside the same run.
        temperature: Option<f64>,
        /// WHICH AGENT this reply belongs to; empty is the process's own
        /// agent. Compaction is a turn taken by the `summarizer` — an ordinary
        /// agent, with its own file and its own conversation — so the reply
        /// must be recorded as its words and never as this agent's answer.
        speaker: String,
    },
    /// Run one tool through its granted capability (Work's single action).
    InvokeTool { tool: ToolId, args_json: String },
    /// Record a fact (I8) beyond what the runtime already logs.
    Emit { kind: EventKind },
    /// Write through StorePort — the agent persists state as data, never
    /// holds a connection (I2).
    Persist { key: String, value_json: String },
    /// Wake me later (heartbeat, retries with backoff).
    Sleep { ms: u64 },
    /// Hand a goal to another agent running in its own Worker (§10 Tier 2,
    /// ADR-008), and take its answer back as an observation. This is the
    /// Python `Tool.from_engine`: the caller never touches the sub-agent's
    /// loop, it sends it a message and waits.
    ///
    /// `batch` is the LINE the call was written on. Calls sharing a batch were
    /// written on one line, which in the Python means "independent, run at the
    /// same time"; the runtime awaits a batch together and the next batch only
    /// afterwards. Increment 05 shipped the ordering half of that rule on a
    /// single-threaded host — one Worker per agent is what makes the
    /// concurrency half real.
    Delegate {
        agent: String,
        goal: String,
        batch: u16,
    },
}

File: packages/agent/src/effect.js
/**
 * EFFECTS — the whole of what a pure step is allowed to WANT.
 *
 * `step` cannot do I/O; it can only describe it, and this closed set is the
 * description (I7). Coarse on purpose: one CallModel, one InvokeTool, never
 * micro-effects, so one turn round is one step in, one effect out, one fact
 * back. Plain serializable data because pending effects must survive a refresh
 * — replay reloads the state and its outstanding effects from the log (I11).
 *
 * FOUR VARIANTS, AND THE FIFTH THAT IS NOT HERE. `docs/PORT-MAP.md` sketches a
 * `Store` effect; the Rust has none and never did — persistence is the driver's
 * business on the way out of `handle`, not something a turn asks for. An effect
 * nothing constructs is a door in the wall of a room nobody enters.
 * @module
 */

/** @typedef {import('@harness/kernel').EndpointName} EndpointName */
/** @typedef {import('@harness/kernel').ToolId} ToolId */
/** @typedef {import('@harness/kernel').Fact} Fact */

/**
 * The assembled paper, opaque here. `packages/context` owns its shape and
 * nothing in this package reads inside one: a Document is built by `assemble`
 * and consumed by the model port, and the loop only carries it between them.
 * @typedef {{sections: unknown[]}} Document
 */

/**
 * How this provider hears the paper — the argument `context.render` takes.
 * It rides the effect because the endpoint and the paper meet in exactly one
 * place, and that place is the only one that can choose a notation (I13).
 * @typedef {{target: 'openai' | 'prose' | 'fragments', vision: boolean, audio: boolean}} ProviderFormat
 */

/**
 * @typedef {(
 *   | {type: 'CallModel', document: Document, format: ProviderFormat, endpoint: EndpointName,
 *      model: string, temperature: number | null, speaker: string}
 *   | {type: 'InvokeTool', tool: ToolId, args: string}
 *   | {type: 'Emit', fact: Fact}
 *   | {type: 'Delegate', agent: string, goal: string, batch: number}
 * )} Effect
 */

/** Every effect type, so a driver can refuse an unknown one by name. */
export const EFFECT_TYPES = /** @type {const} */ (['CallModel', 'InvokeTool', 'Emit', 'Delegate'])

/**
 * Ask the model, with a paper (I13: nothing reaches a model except as a
 * Document — carrying one here is what makes an ad-hoc string unrepresentable).
 *
 * `model` is a CATALOGUE KEY and never a URL: the adapter resolves it and
 * attaches the credential downstream of every grant, so no concrete endpoint
 * exists upstream of the broker (I6). `speaker` says WHICH agent this reply
 * will belong to — empty is the process's own agent — because compaction is a
 * turn taken by the summarizer, and its words must never land as this agent's
 * answer.
 * @param {{document: Document, format: ProviderFormat, endpoint: EndpointName,
 *          model?: string, temperature?: number | null, speaker?: string}} call
 * @returns {Effect}
 */
export function callModel(call) {
  return {
    type: 'CallModel',
    document: call.document,
    format: call.format,
    endpoint: call.endpoint,
    model: call.model ?? '',
    temperature: call.temperature ?? null,
    speaker: call.speaker ?? '',
  }
}

/**
 * Run one tool through its granted capability. `args` is the JSON TEXT the
 * model wrote, not a parsed object: a refusal quotes it back verbatim so the
 * model can see what it actually sent, and parsing here would lose that.
 * @param {ToolId} tool @param {string} args @returns {Effect}
 */
export function invokeTool(tool, args) {
  return { type: 'InvokeTool', tool, args }
}

/** Record a fact (I8) beyond what the driver already logs. @param {Fact} fact @returns {Effect} */
export function emit(fact) {
  return { type: 'Emit', fact }
}

/**
 * Hand a goal to another agent and take its answer back as an observation.
 * The caller never touches the sub-agent's loop; it sends a message and waits.
 *
 * `batch` is the LINE the call was written on. Calls sharing a batch were
 * written on one line, which means "independent, run together": the driver
 * awaits a batch whole and starts the next only afterwards. The index rides out
 * here because the ORDER half of that rule is pure and the CONCURRENCY half —
 * one Worker per agent — belongs to the driver, which needs to be told which
 * calls it may overlap.
 * @param {string} agent @param {string} goal @param {number} batch @returns {Effect}
 */
export function delegate(agent, goal, batch) {
  return { type: 'Delegate', agent, goal, batch }
}
